torch_xla/csrc/dl_convertor.cpp

```diff
@@ -0,0 +1,345 @@
+    return DLDeviceType::kDLCPU;
+  } else if (device.client()->platform_id() == xla::CudaId()) {
+    return DLDeviceType::kDLCUDA;
+  }
+  XLA_ERROR() << "Device " << device.DebugString()
+              << " cannot be used as a DLPack device.";
+}
+
+DLDevice DLDeviceForDevice(const xla::PjRtDevice& device) {
+  DLDevice dlDevice;
+  dlDevice.device_type = DLDeviceTypeForDevice(device);
+  dlDevice.device_id = device.local_hardware_id();
+  return dlDevice;
+}
+
+DLDataType PrimitiveTypeToDLDataType(xla::PrimitiveType type) {
+  switch (type) {
+    case xla::PrimitiveType::S8:
+      return DLDataType{kDLInt, 8, 1};
+    case xla::PrimitiveType::S16:
+      return DLDataType{kDLInt, 16, 1};
+    case xla::PrimitiveType::S32:
+      return DLDataType{kDLInt, 32, 1};
+    case xla::PrimitiveType::S64:
+      return DLDataType{kDLInt, 64, 1};
+    case xla::PrimitiveType::U8:
+      return DLDataType{kDLUInt, 8, 1};
+    case xla::PrimitiveType::U16:
+      return DLDataType{kDLUInt, 16, 1};
+    case xla::PrimitiveType::U32:
+      return DLDataType{kDLUInt, 32, 1};
+    case xla::PrimitiveType::U64:
+      return DLDataType{kDLUInt, 64, 1};
+    case xla::PrimitiveType::F16:
+      return DLDataType{kDLFloat, 16, 1};
+    case xla::PrimitiveType::F32:
+      return DLDataType{kDLFloat, 32, 1};
+    case xla::PrimitiveType::F64:
+      return DLDataType{kDLFloat, 64, 1};
+    case xla::PrimitiveType::BF16:
+      return DLDataType{kDLBfloat, 16, 1};
+    case xla::PrimitiveType::PRED:
+      return DLDataType{kDLBool, 8, 1};
+    case xla::PrimitiveType::C64:
+      return DLDataType{kDLComplex, 64, 1};
+    case xla::PrimitiveType::C128:
+      return DLDataType{kDLComplex, 128, 1};
+    default:
+      XLA_ERROR() << "XLA type " << xla::PrimitiveType_Name(type)
+                  << " has no DLPack equivalent";
+  }
+}
+
+std::vector<int64_t> StridesForShape(xla::PrimitiveType element_type,
+                                     absl::Span<const int64_t> dimensions,
+                                     const xla::Layout& layout) {
+  XLA_CHECK_EQ(dimensions.size(), layout.minor_to_major().size());
+  std::vector<int64_t> strides;
+  strides.resize(dimensions.size());
+  int64_t stride = 1;
+  for (int i : layout.minor_to_major()) {
+    strides[i] = stride;
+    stride *= dimensions[i];
+  }
+  return strides;
+}
+
+// Convert an XLA tensor to a dlPack tensor.
+DLManagedTensor* toDLPack(const at::Tensor& input) {
+  std::shared_ptr<runtime::ComputationClient::Data> handle =
+      get_data_handle(input);
+  XLA_CHECK(handle != nullptr)
+      << "Could not extract a valid data handle from the input tensor";
+
+  std::shared_ptr<xla::PjRtBuffer> pjrt_buffer =
+      runtime::GetComputationClient()->GetPjRtBuffer(handle);
+  XLA_CHECK(pjrt_buffer != nullptr) << "Could not get a valid pjrt_buffer";
+
+  if (pjrt_buffer->IsTuple()) {
+    XLA_ERROR() << "Unimplemented. BufferToDLPackManagedTensor is not "
```

Comment:
micro-nit: use `XLA_CHECK(!pjrt_buffer->IsTuple()) << "message"` here and below

Reply:
done